src/sections/upload-dataset-files/UploadDatasetFiles.tsx

```diff
@@ -1,19 +1,104 @@
-import { useEffect } from 'react'
+import { useEffect, useState } from 'react'
 import { FileRepository } from '../../files/domain/repositories/FileRepository'
 import { useLoading } from '../loading/LoadingContext'
 import { useDataset } from '../dataset/DatasetContext'
 import { PageNotFound } from '../page-not-found/PageNotFound'
 import { BreadcrumbsGenerator } from '../shared/hierarchy/BreadcrumbsGenerator'
+import { FileUploader } from './FileUploader'
+import { FileUploadTools } from '../../files/domain/models/FileUploadState'
+import { uploadFile } from '../../files/domain/useCases/uploadFile'
+import { useTranslation } from 'react-i18next'
 
 interface UploadDatasetFilesProps {
   fileRepository: FileRepository
 }
 
-export const UploadDatasetFiles = ({
-  fileRepository: _fileRepository
-}: UploadDatasetFilesProps) => {
+export const UploadDatasetFiles = ({ fileRepository: fileRepository }: UploadDatasetFilesProps) => {
+  const { t } = useTranslation('files')
   const { setIsLoading } = useLoading()
   const { dataset, isLoading } = useDataset()
+  const [fileUploaderState, setState] = useState(FileUploadTools.createNewState([]))
+  const [uploadingToCancelMap, setUploadingToCancelMap] = useState(new Map<string, () => void>())
+  const [uploadFinished, setUploadFinished] = useState(new Set<string>())
+  const [semaphore, setSemaphore] = useState(new Set<string>())
+
+  const timeout = (delay: number) => new Promise((res) => setTimeout(res, delay))
+  const limit = 6
+
+  const acquireSemaphore = async (file: File) => {
+    const key = FileUploadTools.key(file)
+    setSemaphore((x) => (x.size >= limit ? x : x.add(key)))
+    while (!semaphore.has(key)) {
+      await timeout(500)
```

Comment:
Why do you need to manually delay this process?

Reply:
I could have used here events and listeners, or some other publish-subscribe thing. But it would make the code more complex. I looked for some examples I could find online, but did not like them. Keeping the while loop without delay makes one CPU core go 100%, the delay solves it and the upload feels responsive enough. There might be a better solution for this, I did not find one yet...